src/v/io/io_queue.cc

```diff
@@ -0,0 +1,197 @@
+      ops_, seastar::semaphore::max_counter());
+
+    auto file = std::exchange(file_, nullptr);
+    try {
+        co_await file->close();
+    } catch (...) {
+        log.warn(
+          "Unable to close file {}: {}", path_, std::current_exception());
+    }
+}
+
+void io_queue::submit_read(page& page) noexcept {
+    page.set_flag(page::flags::read);
+    enqueue_pending(page);
+}
+
+void io_queue::submit_write(page& page) noexcept {
+    /*
+     * if the page is not on any list, then in queue it.
+     */
+    if (!page.io_queue_hook.is_linked()) {
+        page.set_flag(page::flags::write);
+        enqueue_pending(page);
+        return;
+    }
+
+    assert(page.test_flag(page::flags::write));
+
+    /*
+     * if it isn't queued, then that means the dispatcher has handled it. in
+     * this case, new data may have arrived into the page. we mark it as queued
+     * as a signal to the I/O completion continuation to requeue the write.
+     */
+    if (!page.test_flag(page::flags::queued)) {
+        log.debug("Marking inflight I/O to be requeued");
+        page.set_flag(page::flags::queued);
+    }
+}
+
+void io_queue::enqueue_pending(page& page) noexcept {
+    pending_.push_back(page);
+    page.set_flag(page::flags::queued);
+    cond_.signal();
+}
+
+seastar::future<> io_queue::dispatch() noexcept {
+    while (true) {
+        co_await cond_.wait([this] { return !pending_.empty() || stop_; });
+
+        if (stop_) {
+            co_return;
+        }
+
+        auto units = co_await seastar::get_units(ops_, 1);
+
+        if (file_ == nullptr) {
+            continue;
+        }
+
+        auto& page = pending_.front();
+        pending_.pop_front();
+        assert(page.test_flag(page::flags::queued));
+        page.clear_flag(page::flags::queued);
+        running_.push_back(page);
+
+        if (page.test_flag(page::flags::read)) {
+            dispatch_read(page, std::move(units));
+
+        } else if (page.test_flag(page::flags::write)) {
+            dispatch_write(page, std::move(units));
+
+        } else {
+            assert(false);
+            std::terminate();
+        }
+    }
+}
+
+void io_queue::dispatch_read(
+  page& page, seastar::semaphore_units<> units) noexcept {
+    log.debug("Reading {} at {}~{}", path_, page.offset(), page.data().size());
+
+    std::ignore = seastar::with_gate(
+      gate_, [this, &page, units = std::move(units)]() mutable {
+          return file_
+            ->dma_read(
+              page.offset(), page.data().get_write(), page.data().size())
+            .then([this, &page, units = std::move(units)](auto) {
+                running_.erase(request_list_type::s_iterator_to(page));
+                if (complete_) {
+                    complete_(page);
+                }
+            })
+            .handle_exception([this](std::exception_ptr eptr) {
+                log.error("Read failed to {}: {}", path_, eptr);
+                std::terminate();
```

Comment:
nit: might be useful to add a TODO comment for the future and same below :)

Reply:
What is there todo differently? I'm being facetious, but it's actually a reasonable behavior in that it's how we handle I/O errors in Redpanda today. In reality an explicit goal of the I/O layer is to offer finer grained handling of I/O errors like ENOSPC, but that isn't really a TODO so much as another phase of the project.

Reply:
Are there never transient issues, like EAGAIN or ECANCELED or something?

Reply:
Changed to vassert per feedback on previous comment with terminate/assert.

Reply:
> Are there never transient issues, like EAGAIN or ECANCELED or something?

There may be. With Redpanda's hard fail for I/O errors, I've only ever seen ENOSPC. I think that interrupted system calls might be retried automatically. When we get rid of the hard fail for all IO errors we'll have to revisit this.

Reply:
> With Redpanda's hard fail for I/O errors

Oh do we do this today? That's news to me.

Reply:
>Oh do we do this today? That's news to me.

yeh, see segment_appender.cc. reads handled by seastar in input stream may have more error handling.